File: Python2/DupeChecker.py
#!/usr/bin/python2.7
# dupechecker
# Finds duplicate files based on md5 and filesize
# Created by LAMMJohnson for the gentoomen 4chantoolbox project

import sys, getopt, os, hashlib

## Defaults
recursive   = False       #defaults to not recursive
directory   = "./"        #default is current dir
ask         = False	  #defaults to requiring confirmation
defChoice   = 'yes'	  #used when ask is False
dummy       = False

# All important list of files
filestats   = []

class fileblock():
    def __init__(self, inpath):
        self.wholepath = inpath
        self.size = os.path.getsize(inpath)
        self.md5 = ""

    def get_md5(self):
        print "Generating md5 for " + self.wholepath
        self.md5 = md5_for_file(self.wholepath)

def add_dir(dirpath):
    print "Scanning directory " + dirpath
    for tempfile in os.listdir(dirpath):
        if os.path.isdir(dirpath + tempfile) and recursive:
            add_dir(dirpath + tempfile + "/")
        elif os.path.isfile(dirpath + tempfile):
            filestats.append( fileblock( dirpath + tempfile ) )

# get user input to determine if it's deletion time or not
def get_choice():
    choose = ""
    validOptions = ( 'y', 'n', 'o', 'b', 'yes', 'no', 'other', 'both')
    while not choose in validOptions :
        choose = raw_input('(y)es / (n)o / the (o)ther file / (b)oth ? ')
    return choose

def md5_for_file(filename, block_size=2**20):
    f = open(filename)
    md5 = hashlib.md5()
    while True:
        data = f.read(block_size)
        if not data:
            break
        md5.update(data)
    return md5.digest()

def pretty_size(size):
    suffixes = [("B",2**10), ("K",2**20), ("M",2**30), ("G",2**40), ("T",2**50)]
    for suf, lim in suffixes:
        if size > lim:
            continue
        else:
            return round(size/float(lim/2**10),2).__str__()+suf

def remfile(infile):
    if os.path.isfile(infile.wholepath):
        os.remove(infile.wholepath)
        print infile.wholepath + " deleted."

def showmatch(filea, fileb):
    print "=========================== Size Match ==========================="
    print "File:    " + filea.wholepath
    print "Matches: " + fileb.wholepath
    print "Size:    " + pretty_size( filea.size )

def usage():
    print "=================================================================="
    print "Usage: Dupechecker-Python [OPTION] <directory>"
    print "Script to remove duplicate files"
    print " "
    print " -d/--dummy        No deletion. Output findings without prompting"
    print " -r/--recursive    Recurse through directories"
    print " -y/--yes-to-all   Remove duplicates without prompting"
    print " -h/--help         This help text"
    print "=================================================================="
    sys.exit(' ')

## Rip out any genuine paths in the args
for arg in sys.argv:
    if os.path.isdir(arg):
        directory = arg

## Handle Args
try:
    opts, args = getopt.gnu_getopt( sys.argv, "dhry", ["dummy", "help", "recursive", "yes-to-all"] )

except getopt.GetoptError:
    print str(err)
    usage()

for o, a in opts:
    if o in ("-d", "--dummy"):
        dummy = True
    elif o in ("-r", "--recursive"):
        recursive = True
    elif o in ("-y", "--yes-to-all"):
        ask = False
    elif o in ("-h", "--help"):
        usage()

# Spit out a warning if we're using the current dir
if directory == "./":
    print "No valid directory given. Using current directory."

# If the dir doesn't end with a "/" we need to add one
directory = os.path.join(directory, '')

## Main part
#Get a list of files only (no dirs)
add_dir(directory)

#to allow us to catch exceptions
# The real meat of it -- compare listed files.
for i in range(len(filestats)):
    startfile = filestats[i]
    # be careful to check we haven't deleted either the file in a previous action
    if os.path.isfile(startfile.wholepath):

        for e in range(i + 1, len(filestats)):
            checkagainst = filestats[e]
            # be careful to check we haven't deleted either the file in a previous action
            if os.path.isfile(checkagainst.wholepath):

                if startfile.size == checkagainst.size:
                    showmatch(startfile, checkagainst)

                    # Generate md5s if necessary
                    for f in (startfile, checkagainst):
                        if f.md5 == "":
                            f.get_md5()
                
                    if startfile.md5 == checkagainst.md5:
                        if dummy:
                            print "=== Full md5 match found ==="

                        else:
                            print "Match found! Would you like to delete duplicate file " + startfile.wholepath + ""
    
                            # Allowance for the automatic overwrite flag
                            if ask:
                                choice = get_choice()
                            else:
				                choice = defChoice

                            # In instructed to, delete the file    
                            if choice in ('o', 'other', 'b', 'both'):
                                remfile(checkagainst)
                            if choice in ('y', 'yes', 'b', 'both') or not ask:
                                remfile(startfile)
                                break
                    else:
                        print "No md5 match."

File: Ruby/Dupechecker.rb
#!/usr/bin/ruby
# dupechecker
# Finds duplicate files based on md5 and filesize
# Created by DukeDerpington for the gentoomen 4chantoolbox project
#             (i also created the scraper :P just sayin')

require "getoptlong"
require "digest/md5"

## Defaults
@recursive   = false       #defaults to not recursive
@directory   = "./"        #default is current dir
@ask         = true        #defaults to requiring confirmation
@dummy       = false

# All important list of files
@filestats   = []

class Fileblock
    attr_reader :wholepath, :size, :md5
    def initialize inpath
        @wholepath = inpath
        @size = File.size(inpath)
        @md5 = ""
    end
    def get_md5
        @md5 = Digest::MD5.hexdigest(IO.read(@wholepath))
    end
end

def add_dir(dirpath)
    puts "Scanning directory " + dirpath
    (Dir.entries(dirpath) - ". ..".split).each do |tempfile|
        if File.directory? dirpath + tempfile and @recursive then
            add_dir(dirpath + tempfile + "/")
        elsif File.file? dirpath + tempfile then
            @filestats << Fileblock.new(dirpath + tempfile)
        end
    end
end

# get user input to determine if it's deleteion time or not
def get_choice
    choose = ""
    validOptions = ['y', 'n', 'o', 'b', 'yes', 'no', 'other', 'both']
    while not validOptions.include? choose do
        print ('(y)es / (n)o / the (o)ther file / (b)oth ? ')
        choose = gets.chomp
    end
    return choose
end

def pretty_size size
    suffixes = { "B" => 2**10, "K" => 2**20, "M" => 2**30, "G" => 2**40, "T" => 2**50 }
    suffixes.each do |suf, lim|
        if size > lim then
            next
        else
            return (size/lim/2**10.0).to_i.to_s + suf
        end
    end
end

def remfile infile
    if File.file?(infile.wholepath) then
        File.delete(infile.wholepath)
        puts infile.wholepath + " deleted."
    end
end

def showmatch filea, fileb
    puts "=========================== Size Match ==========================="
    puts "File:    " + filea.wholepath
    puts "Matches: " + fileb.wholepath
    puts "Size:    " + pretty_size( filea.size )
end

def usage
    puts "=================================================================="
    puts "Usage: Dupechecker-Ruby [OPTION] <directory>"
    puts "Script to remove duplicate files"
    puts " "
    puts " -d/--dummy        No deletion. Output findings without prompting"
    puts " -r/--recursive    Recurse through directories"
    puts " -y/--yes-to-all   Remove duplicates without prompting"
    puts " -h/--help         This help text"
    puts "=================================================================="
    exit
end

## Rip out any genuine paths in the args
ARGV.each do |arg|
    if File.directory? arg then
        @directory = arg
    end
end

## Handle Args
opts = GetoptLong.new(
    [ '-d', "--dummy", GetoptLong::NO_ARGUMENT ],
    [ '--help', "-h", GetoptLong::NO_ARGUMENT ],
    [ '--recursive', "-r", GetoptLong::NO_ARGUMENT ],
    [ '--yes-to-all', "-y", GetoptLong::NO_ARGUMENT ]
)

opts.each do |opt, arg|
    if ['-d', '--dummy'].include? opt then
        @dummy = true
    elsif ['-h', '--help'].include? opt then
        usage()
    elsif ['-r', '--recursive'].include? opt then
        @recursive = true
    elsif ['-y', '--yes-to-all'].include? opt then
        @ask = false
    end
end

# Spit out a warnign if we're using the current dir
if @directory == "./" then
    puts "No valid directory given. Using current directory."
end

# If the dir doesn't end with a "/" we need to add one
if not @directory[ - 1, 1 ] == "/" then
    @directory += "/"
end

## Main part
#Get a list of files only (no dirs)
add_dir(@directory)

#to allow us to catch exceptions
begin
    # The real meat of it -- compare listed files.
    @filestats.length.times do |i|
        startfile = @filestats[i]
        # be careful to check we haven't deleted either the file in a previous action
        if File.file?(startfile.wholepath) then

            (i+1).upto @filestats.length-1 do |e|
                checkagainst = @filestats[e]
                # be careful to check we haven't deleted either the file in a previous action
                if not File.file?(checkagainst.wholepath) then
                    break
                end

                if startfile.size == checkagainst.size then
                    showmatch(startfile, checkagainst)

                    # Generate md5s if necessary
                    [startfile, checkagainst].each do |f|
                        f.get_md5() if f.md5 == ""
                    end
                
                    if @dummy then
                        puts "=== Full md5 match found ==="
                    else
                        if startfile.md5 == checkagainst.md5 then
                            puts "Match found! Would you like to delete duplicate file " + startfile.wholepath.inspect + ""
    
                            # Allowance for the automatic overwrite flag
                            choice = get_choice() if @ask

                            # In instructed to, delete the file    
                            if ['o', 'other', 'b', 'both'].include? choice then
                                remfile(checkagainst)
                            end

                            if ['y', 'yes', 'b', 'both'].include? choice or not @ask then
                                remfile(startfile)
                                break
                            end

                        else
                            puts "No md5 match."
                        end
                    end
                end
            end
        end
    end
rescue Interrupt, SystemExit
    puts "Interrupted. Excited cleanly."
    exit
end
